### PR TITLE
feat(components/containers): allow to pass custom renderers to List

### PR DESCRIPTION
**Pull request by romainseb**
**What is the problem this PR is trying to solve?**
We have a static dictionary for the cell renderers

**What is the chosen solution to this problem?**
Allow to pass new cell rederers via props

**Please check if the PR fulfills these requirements**

* [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
* [x] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
* [x] Docs have been added / updated (for bug fixes / features)
* [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/components/src/List/ListToVirtualizedList/ListToVirtualizedList.component.js [hunk 1]
```diff
@@ -111,6 +114,7 @@ ListToVirtualizedList.propTypes = {
 };
 ListToVirtualizedList.defaultProps = {
 	displayMode: 'table',
+	cellDictionary: {},
```
Comment (jsomsanith) [inline packages/components/src/List/ListToVirtualizedList/ListToVirtualizedList.component.js]: This is not necessary, you only spread it, and spreading undefined is supported.

File: packages/components/src/List/ListToVirtualizedList/ListToVirtualizedList.test.js [hunk 2]
```diff
@@ -118,6 +119,23 @@ describe('ListToVirtualizedList', () => {
 		});
 	});
 
+	it('should support custom cell renderer', () => {
+		const clonedProps = cloneDeep(props);
```
Comment (jsomsanith) [inline packages/components/src/List/ListToVirtualizedList/ListToVirtualizedList.test.js]: You don't need to deep clone the props, just spread the props, and then add your `cellDictionary` props in your component.
```
<ListToVirtualizedList {...props} cellDictionnary={customDictionnary} />
```

File: packages/components/src/List/ListToVirtualizedList/ListToVirtualizedList.test.js [hunk 3]
```diff
@@ -118,6 +119,23 @@ describe('ListToVirtualizedList', () => {
...
+		const renderer = function test() {
+			return 'ok';
+		};
+		clonedProps.cellDictionary = { customType: { cellRenderer: renderer } };
+		const wrapper = shallow(<ListToVirtualizedList {...clonedProps} />);
+
+		// then
+		wrapper.find(VirtualizedList.Content).forEach(element => {
```
Comment (jsomsanith) [inline packages/components/src/List/ListToVirtualizedList/ListToVirtualizedList.test.js]: Insteed of going through all the columns, `Array.find()` the wanted column and test the assertion on it

File: packages/components/src/List/List.md [hunk 4]
```diff
@@ -2,40 +2,40 @@
...
-| items | Array of object to display |
-| titleProps | object to control the title cell |
-| itemProps | |
+| name           | description                                                               |
+| -------------- | ------------------------------------------------------------------------- |
+|  columns       |  Array of {key, label, type} define which columns to display              |
+| items          | Array of object to display                                                |
+|  titleProps    |  object to control the title cell                                         |
+| itemProps      |                                                                           |
+| cellDictionary | Array of object to define custom cell renderers or override existing ones |
```
Comment (jsomsanith) [inline packages/components/src/List/List.md]: It's not an array of object, but an object with key/value, key = renderer name to use in columns props, value = the renderer

File: packages/components/stories/List.js [hunk 5]
```diff
@@ -9,6 +10,22 @@ import { cloneDeep } from 'lodash';
 import { List, IconsProvider } from '../src/index';
 import i18n, { LanguageSwitcher } from './config/i18n';
 
+/**
+ * Cell renderer that displays hello + text
+ */
+function CellWithHello({ cellData }) {
+	return (
+		<div>
+			<div>hello {cellData} !</div>
```
Comment (jsomsanith) [inline packages/components/stories/List.js]: div in div to add a string :)

File: packages/components/stories/List.js [hunk 6]
```diff
@@ -584,6 +605,30 @@ storiesOf('List', module)
...
+
+		customProps.list.items = itemsForListWithIcons;
+		customProps.list.cellDictionary = { hello: { cellRenderer: CellWithHello } };
+
+		return (
+			<div style={{ height: '60vh' }} className="virtualized-list">
+				<h1>List</h1>
+				<p>
+					Display the list in table mode.<br />
+					This is the default mode.
```
Comment (jsomsanith) [inline packages/components/stories/List.js]: Please update the description explaining what this story shows

File: packages/components/stories/List.js (hunk 5 above)
Comment (romainseb) [reply packages/components/stories/List.js]: i can add another :D

File: packages/containers/src/List/List.container.js [hunk 7]
```diff
@@ -203,6 +204,18 @@ class List extends React.Component {
 			}
 		}
 
+		if (props.cellDictionary) {
+			props.list.cellDictionary = Object.keys(props.cellDictionary).reduce((acc, key) => {
+				const current = props.cellDictionary[key];
+				// eslint-disable-next-line no-param-reassign
+				acc[key] = {
+					...omit(current, ['cellRendererComponent', 'cellRendererComponentId']),
```
Comment (jsomsanith) [inline packages/containers/src/List/List.container.js]: `cellRendererComponentId` is not used here. Why do we have it ?

File: output/components.eslint.txt [hunk 8]
```diff
@@ -18,6 +18,9 @@ The react/require-extension rule is deprecated. Please use the import/extensions
 /home/travis/build/Talend/ui/packages/components/src/JSONSchemaRenderer/JSONSchemaRenderer.component.js
   76:56  error  'properties' is defined but never used  no-unused-vars
 
+/home/travis/build/Talend/ui/packages/components/src/List/ListToVirtualizedList/ListToVirtualizedList.test.js
+  3:10  error  'cloneDeep' is defined but never used  no-unused-vars
```
Comment (jsomsanith) [inline output/components.eslint.txt]: You forgot to remove the import.

File: packages/containers/src/List/List.container.js (hunk 7 above)
Comment (romainseb) [reply packages/containers/src/List/List.container.js]: Just forgot it on the next line 🙄

File: packages/containers/src/List/List.container.js [hunk 9]
```diff
@@ -203,6 +204,18 @@ class List extends React.Component {
 			}
 		}
 
+		if (props.cellDictionary) {
+			props.list.cellDictionary = Object.keys(props.cellDictionary).reduce((acc, key) => {
+				const current = props.cellDictionary[key];
+				// eslint-disable-next-line no-param-reassign
+				acc[key] = {
+					...omit(current, ['cellRendererComponent']),
+					cellRenderer: props.getComponent(current.cellRendererComponent),
```
Comment (jmfrancois) [inline packages/containers/src/List/List.container.js]: I would go to just component.
`props.getComponent(current.component)`

**Pull request merged**